infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,228 @@
+     * @param retries              The number of retries.
+     * @return A result of reading records.
+     */
+    @VisibleForTesting
+    CompletableFuture<ReadBatch> readRecords(TransferBatchRequest transferBatchRequest, int retries) {
+        return CompletableFuture
+                .supplyAsync(() -> {
+                            Supplier<Map<Long, ILogData>> readSupplier = () -> addressSpaceView
+                                    .simpleProtocolRead(transferBatchRequest.getAddresses(), readOptions);
+
+                            return Result
+                                    .of(readSupplier)
+                                    .mapError(StateTransferException::new);
+                        }
+                ).thenApply(readResult ->
+                        readResult.map(records -> checkReadRecords(
+                                transferBatchRequest.getAddresses(),
+                                records,
+                                transferBatchRequest.getDestination()
+                        ))
+                ).thenCompose(checkedReadResult -> {
+                    // If the protocol read threw an exception, we retry.
+                    if (checkedReadResult.isError()) {
+                        return retryReadRecords(transferBatchRequest, retries);
+                    } else {
+                        // If the protocol read went fine but we read only part of the records,
+                        // we retry.
+                        ReadBatch readBatch = checkedReadResult.get();
+                        if (readBatch.getStatus() == ReadBatch.ReadStatus.FAILED) {
+                            return retryReadRecords(readBatch.createRequest(), retries);
+                        } else {
+                            return CompletableFuture.completedFuture(readBatch);
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Handle read errors. Exponential backoff policy is utilized.
+     *
+     * @param transferBatchRequest TransferBatchRequest of addresses to be read.
+     * @param retriesTried         Configurable number of retries for the current transferBatchRequest.
+     * @return Future of the result of the retry.
+     */
+    @VisibleForTesting
+    CompletableFuture<ReadBatch> retryReadRecords
+    (TransferBatchRequest transferBatchRequest, int retriesTried) {
+        try {
+            AtomicInteger readRetries = new AtomicInteger(retriesTried);
+            return IRetry.build(ExponentialBackoffRetry.class, RetryExhaustedException.class, () -> {
+                if (readRetries.incrementAndGet() >= maxReadRetries) {
+                    throw new RetryExhaustedException("Read retries are exhausted");
+                }
+
+                // If a pipeline completed exceptionally, than return the error.
+                CompletableFuture<ReadBatch> pipelineFuture =
+                        readRecords(transferBatchRequest, readRetries.get())
```

Comment:
I would not necessary agree about readRecords and retryReadRecords because it is a very common practice in declarative programming for the functions to call each other in the retry scenarios, since they compose much easier in this way.
Regarding exception handling I understand that this is not the cleanest way. The problem is that we want the exponential back off on read failures while the AddressSpaceView provides us with just simple retries for all the cases, so it's preferable to handle them here. 
The next PR was supposed to handle the WrongEpochException differently and not handle it in the retry loop. I also don't think I will be able to rewrite this block of code today, I think it would be better to address these specific improvements in a separate PR, as this one is more concerned with changing a clustering model.